python/modules/qcdSmearProducer.py

```diff
@@ -0,0 +1,258 @@
+#!/usr/bin/env python
+import os, sys
+import ROOT
+ROOT.PyConfig.IgnoreCommandLineOptions = True
+from importlib import import_module
+
+from PhysicsTools.NanoAODTools.postprocessing.framework.datamodel import Collection, Object
+from PhysicsTools.NanoAODTools.postprocessing.framework.eventloop import Module
+from PhysicsTools.NanoAODTools.postprocessing.tools import deltaPhi, deltaR, closest
+
+class qcdSmearProducer(Module): 
+    def __init__(self):
+        self.writeHistFile=True
+        self.metBranchName="MET"
+        self.xBinWidth = 0.01
+        self.minWindow = 0.01
+        self.maxWindow = 0.5
+        self.nSmears = 100
+        self.nSmearJets = 2
+        self.nBootstraps = 50
+        self.LINEAR_GRANULATED=True
+        self.winType = self.LINEAR_GRANULATED
+        self.doFlatSampling = True
+        self.canSmear  =True
+        self.respInputName = "JetResByFlav"
+        self.respFileName = "file:/eos/uscms/store/user/ddash/qcd_smeared/resTailOut_combined_filtered_CHEF_puWeight_weight_WoH_NORMALIZED.root"
+      #  self.respHistoName = self.ptmapping(jets)
+      #  self.targeth = self.loadHisto(self.respFileName,self.respHistoName)
+        
+    def loadHisto(self,filename,hname):
+        tf = ROOT.TFile.Open(filename)
+        hist = tf.Get(hname)
+        hist.SetDirectory(None)
+        tf.Close()
+        return hist
+    
+    def ptmapping(self,jets):
+        #self.jetspt      = self.analyze(event).jets
+        self.ptrange = [jets.pt.range(0,50),jets.pt.range(50,75),jets.pt.range(75,100),jets.pt.range(100,125),jets.pt.range(125,150),jets.pt.range(150,200),jets.pt.range(200,250),jets.pt.range(300,400),jets.pt.range(400,500),jets.pt.range(500,700),jets.pt.range(700,1000),jets.pt.range(1000,1500),jets.pt.range(1500,4000)]
```

Comment:
Not familiar with this function. What does range do? Is this tested?

Reply:
It is a user-defined function. Response curves are saved in different pt ranges so the pt range wise the corresponding curve is used.